src/mechanics/IGA/NURBS.h

```diff
@@ -0,0 +1,540 @@
+                                                                int degree, const std::vector<double>& knots,
+                                                                const std::vector<double>& weights)
+    {
+        assert(derivativeOrder >= 0 || derivativeOrder <= 2);
+
+        Eigen::MatrixXd ders = BasisFunctionsAndDerivatives(derivativeOrder, parameter, spanIdx, degree, knots);
+
+        // NURBS specific ...
+        Eigen::VectorXd sum(derivativeOrder + 1);
+        sum.setZero(derivativeOrder + 1);
+
+        for (int i = 0; i <= degree; i++)
+        {
+            double weight = weights[spanIdx - degree + i];
+            if (derivativeOrder == 0)
+                sum(0) += ders(0, i) * weight;
+            else if (derivativeOrder == 1)
+            {
+                sum(0) += ders(0, i) * weight;
+                sum(1) += ders(1, i) * weight;
+            }
+            else
+            {
+                sum(0) += ders(0, i) * weight;
+                sum(1) += ders(1, i) * weight;
+                sum(2) += ders(2, i) * weight;
+            }
+        }
+
+        Eigen::VectorXd basisFctsDerivativesRational(degree + 1);
+        basisFctsDerivativesRational.setZero(degree + 1);
+
+        for (int i = 0; i <= degree; i++)
+        {
+            double weight = weights[spanIdx - degree + i];
+            if (derivativeOrder == 0)
+                basisFctsDerivativesRational(i) = ders(0, i) * weight / sum(0);
+            else if (derivativeOrder == 1)
+                basisFctsDerivativesRational(i) =
+                        (ders(1, i) * sum(0) - ders(0, i) * sum(1)) * weight / (sum(0) * sum(0));
+            else
+            {
+                double sum2 = sum(0) * sum(0);
+                basisFctsDerivativesRational(i) =
+                        weight * (ders(2, i) / sum(0) - 2 * ders(1, i) * sum(1) / (sum2)-ders(0, i) * sum(2) / (sum2) +
+                                  2 * ders(0, i) * sum(1) * sum(1) / (sum2 * sum(0)));
+            }
+        }
+
+        return basisFctsDerivativesRational;
+    }
+
+    //! @brief ... calculates the knot span to given parameters
+    //! @param parameter ... parameter(s) at which the functions are calculated
+    //! @return std::array ... knot span(s) to given parameter(s)
+    const std::array<int, TDimParameter> FindSpan(const Eigen::Matrix<double, TDimParameter, 1>& parameter) const
+    {
+        std::array<int, TDimParameter> parameterIDs;
+        for (int i = 0; i < parameter.rows(); i++)
+        {
+            parameterIDs[i] = FindSpan(parameter[i], mDegree[i], mKnots[i]);
+        }
+        return parameterIDs;
+    }
+
+    Eigen::MatrixXd BasisFunctionsAndDerivativesRational(int der,
+                                                         const Eigen::Matrix<double, TDimParameter, 1>& parameter) const
+    {
+        throw new std::string("Not implemented for an arbitrary dimension!");
```

Comment:
Like @Psirus said, please throw a NuTo::Exception instead. And you could be more specific:
~~~cpp
throw NuTo::Exception(__PRETTY_FUNCTION, "Only implemented for TDim = 1,2. "); // or something like that
~~~

 But I like to point out another bug/error/... . A raw `new` should never appear in our code.